pytest.ini

```diff
@@ -23,9 +23,9 @@ addopts =
   -p pytest_cov
 
   # `pytest-cov`:
-  --cov
```

Comment:
FTR, `pytest-cov` usually behaves more correctly with this setting and most of the config happening in coverage.py. Besides, the idea is that if any other tools use coverage.py directly, they'd be able to rely on the same config. This is the principle I apply to all the integrated tooling.

I'd like this pattern to remain in place.

Reply:
I'm going to revert all the unneeded changes once I get things working (well.. if I get things working)

Reply:
Okay. Just postpone merging since I synchronize the configs to have the same shape across repos, so I want to make sure the end result fits within the structure I got for that.

Reply:
Unfortunately I can't get it to work without this change

Every time I do it changes the package back to `"."`

Reply:
I've gone so far as to dump the internal coverage options and make sure they are the same and it still changes it back to `"."`

Reply:
Working
```
['_abc_impl', '->', <_abc._abc_data object at 0x105494140>]
['_config_contents', '->', b'[paths]\nsource =\n  yarl\n  tests\n  */lib/pypy*/site-packages/yarl\n  */lib/python*/site-packages/yarl\n  */Lib/site-packages/yarl\n\n[report]\nskip_covered = true\nskip_empty = true\nshow_missing = true\n\n[run]\nbranch = true\ncover_pylib = false\nplugins = Cython.Coverage\nsource_pkgs =\n  tests\n  yarl\n']
['_crash', '->', None]
['_include', '->', None]
['_is_protocol', '->', False]
['_is_runtime_protocol', '->', False]
['_omit', '->', None]
['_set_attr_from_config_option', '->', <bound method CoverageConfig._set_attr_from_config_option of <coverage.config.CoverageConfig object at 0x10544f7d0>>]
['branch', '->', True]
['command_line', '->', None]
['concurrency', '->', []]
['config_file', '->', '/Users/bdraco/yarl/.coveragerc']
['config_files_attempted', '->', ['/Users/bdraco/yarl/.coveragerc']]
['config_files_read', '->', ['/Users/bdraco/yarl/.coveragerc']]
['context', '->', None]
['copy', '->', <bound method CoverageConfig.copy of <coverage.config.CoverageConfig object at 0x10544f7d0>>]
['cover_pylib', '->', False]
['data_file', '->', '/Users/bdraco/yarl/.coverage']
['debug', '->', []]
['debug_file', '->', None]
['debug_info', '->', <bound method CoverageConfig.debug_info of <coverage.config.CoverageConfig object at 0x10544f7d0>>]
['disable_warnings', '->', []]
['dynamic_context', '->', None]
['exclude_also', '->', []]
['exclude_list', '->', ['#\\s*(pragma|PRAGMA)[:\\s]?\\s*(no|NO)\\s*(cover|COVER)']]
['extra_css', '->', None]
['fail_under', '->', 0.0]
['format', '->', None]
['from_args', '->', <bound method CoverageConfig.from_args of <coverage.config.CoverageConfig object at 0x10544f7d0>>]
['from_file', '->', <bound method CoverageConfig.from_file of <coverage.config.CoverageConfig object at 0x10544f7d0>>]
['get_option', '->', <bound method CoverageConfig.get_option of <coverage.config.CoverageConfig object at 0x10544f7d0>>]
['get_plugin_options', '->', <bound method CoverageConfig.get_plugin_options of <coverage.config.CoverageConfig object at 0x10544f7d0>>]
['html_dir', '->', 'htmlcov']
['html_skip_covered', '->', None]
['html_skip_empty', '->', None]
['html_title', '->', 'Coverage report']
['ignore_errors', '->', False]
['include_namespace_packages', '->', False]
['json_output', '->', 'coverage.json']
['json_pretty_print', '->', False]
['json_show_contexts', '->', False]
['lcov_output', '->', 'coverage.lcov']
['parallel', '->', True]
['partial_always_list', '->', ['while (True|1|False|0):', 'if (True|1|False|0):']]
['partial_list', '->', ['#\\s*(pragma|PRAGMA)[:\\s]?\\s*(no|NO)\\s*(branch|BRANCH)']]
['paths', '->', {'source': ['yarl', 'tests', '*/lib/pypy*/site-packages/yarl', '*/lib/python*/site-packages/yarl', '*/Lib/site-packages/yarl']}]
['plugin_options', '->', {}]
['plugins', '->', ['Cython.Coverage']]
['post_process', '->', <bound method CoverageConfig.post_process of <coverage.config.CoverageConfig object at 0x10544f7d0>>]
['post_process_file', '->', <bound method CoverageConfig.post_process_file of <coverage.config.CoverageConfig object at 0x10544f7d0>>]
['precision', '->', 0]
['relative_files', '->', False]
['report_contexts', '->', None]
['report_include', '->', None]
['report_omit', '->', None]
['run_include', '->', []]
['run_omit', '->', []]
['set_option', '->', <bound method CoverageConfig.set_option of <coverage.config.CoverageConfig object at 0x10544f7d0>>]
['show_contexts', '->', False]
['show_missing', '->', True]
['sigterm', '->', False]
['skip_covered', '->', True]
['skip_empty', '->', True]
['sort', '->', None]
['source', '->', ['yarl', 'tests']]
['source_pkgs', '->', ['tests', 'yarl']]
['timid', '->', False]
['xml_output', '->', 'coverage.xml']
['xml_package_depth', '->', 99]

```

Reply:
Not working
```
['_config_contents', '->', b'[paths]\nsource =\n  yarl\n  tests\n  */lib/pypy*/site-packages/yarl\n  */lib/python*/site-packages/yarl\n  */Lib/site-packages/yarl\n\n[report]\nskip_covered = true\nskip_empty = true\nshow_missing = true\n\n[run]\nbranch = true\ncover_pylib = false\nplugins = Cython.Coverage\nparallel=True\nsource = \n  yarl\n  tests\nsource_pkgs =\n  tests\n  yarl\n']
['_crash', '->', None]
['_include', '->', None]
['_is_protocol', '->', False]
['_is_runtime_protocol', '->', False]
['_omit', '->', None]
['_set_attr_from_config_option', '->', <bound method CoverageConfig._set_attr_from_config_option of <coverage.config.CoverageConfig object at 0x1065a5c70>>]
['branch', '->', True]
['command_line', '->', None]
['concurrency', '->', []]
['config_file', '->', '/Users/bdraco/yarl/.coveragerc']
['config_files_attempted', '->', ['/Users/bdraco/yarl/.coveragerc']]
['config_files_read', '->', ['/Users/bdraco/yarl/.coveragerc']]
['context', '->', None]
['copy', '->', <bound method CoverageConfig.copy of <coverage.config.CoverageConfig object at 0x1065a5c70>>]
['cover_pylib', '->', False]
['data_file', '->', '/Users/bdraco/yarl/.coverage']
['debug', '->', []]
['debug_file', '->', None]
['debug_info', '->', <bound method CoverageConfig.debug_info of <coverage.config.CoverageConfig object at 0x1065a5c70>>]
['disable_warnings', '->', []]
['dynamic_context', '->', None]
['exclude_also', '->', []]
['exclude_list', '->', ['#\\s*(pragma|PRAGMA)[:\\s]?\\s*(no|NO)\\s*(cover|COVER)']]
['extra_css', '->', None]
['fail_under', '->', 0.0]
['format', '->', None]
['from_args', '->', <bound method CoverageConfig.from_args of <coverage.config.CoverageConfig object at 0x1065a5c70>>]
['from_file', '->', <bound method CoverageConfig.from_file of <coverage.config.CoverageConfig object at 0x1065a5c70>>]
['get_option', '->', <bound method CoverageConfig.get_option of <coverage.config.CoverageConfig object at 0x1065a5c70>>]
['get_plugin_options', '->', <bound method CoverageConfig.get_plugin_options of <coverage.config.CoverageConfig object at 0x1065a5c70>>]
['html_dir', '->', 'htmlcov']
['html_skip_covered', '->', None]
['html_skip_empty', '->', None]
['html_title', '->', 'Coverage report']
['ignore_errors', '->', False]
['include_namespace_packages', '->', False]
['json_output', '->', 'coverage.json']
['json_pretty_print', '->', False]
['json_show_contexts', '->', False]
['lcov_output', '->', 'coverage.lcov']
['parallel', '->', True]
['partial_always_list', '->', ['while (True|1|False|0):', 'if (True|1|False|0):']]
['partial_list', '->', ['#\\s*(pragma|PRAGMA)[:\\s]?\\s*(no|NO)\\s*(branch|BRANCH)']]
['paths', '->', {'source': ['yarl', 'tests', '*/lib/pypy*/site-packages/yarl', '*/lib/python*/site-packages/yarl', '*/Lib/site-packages/yarl']}]
['plugin_options', '->', {}]
['plugins', '->', ['Cython.Coverage']]
['post_process', '->', <bound method CoverageConfig.post_process of <coverage.config.CoverageConfig object at 0x1065a5c70>>]
['post_process_file', '->', <bound method CoverageConfig.post_process_file of <coverage.config.CoverageConfig object at 0x1065a5c70>>]
['precision', '->', 0]
['relative_files', '->', False]
['report_contexts', '->', None]
['report_include', '->', None]
['report_omit', '->', None]
['run_include', '->', []]
['run_omit', '->', []]
['set_option', '->', <bound method CoverageConfig.set_option of <coverage.config.CoverageConfig object at 0x1065a5c70>>]
['show_contexts', '->', False]
['show_missing', '->', True]
['sigterm', '->', False]
['skip_covered', '->', True]
['skip_empty', '->', True]
['sort', '->', None]
['source', '->', ['yarl', 'tests']]
['source_pkgs', '->', ['tests', 'yarl']]
['timid', '->', False]
['xml_output', '->', 'coverage.xml']
['xml_package_depth', '->', 99]
```

Reply:
```diff
--- not_working.txt	2024-08-31 14:39:29
+++ working.txt	2024-08-31 14:36:26
@@ -1,10 +1,11 @@
-['_config_contents', '->', b'[paths]\nsource =\n  yarl\n  tests\n  */lib/pypy*/site-packages/yarl\n  */lib/python*/site-packages/yarl\n  */Lib/site-packages/yarl\n\n[report]\nskip_covered = true\nskip_empty = true\nshow_missing = true\n\n[run]\nbranch = true\ncover_pylib = false\nplugins = Cython.Coverage\nparallel=True\nsource = \n  yarl\n  tests\nsource_pkgs =\n  tests\n  yarl\n']
+['_abc_impl', '->', <_abc._abc_data object at 0x105494140>]
+['_config_contents', '->', b'[paths]\nsource =\n  yarl\n  tests\n  */lib/pypy*/site-packages/yarl\n  */lib/python*/site-packages/yarl\n  */Lib/site-packages/yarl\n\n[report]\nskip_covered = true\nskip_empty = true\nshow_missing = true\n\n[run]\nbranch = true\ncover_pylib = false\nplugins = Cython.Coverage\nsource_pkgs =\n  tests\n  yarl\n']
 ['_crash', '->', None]
 ['_include', '->', None]
 ['_is_protocol', '->', False]
 ['_is_runtime_protocol', '->', False]
 ['_omit', '->', None]
-['_set_attr_from_config_option', '->', <bound method CoverageConfig._set_attr_from_config_option of <coverage.config.CoverageConfig object at 0x1065a5c70>>]
+['_set_attr_from_config_option', '->', <bound method CoverageConfig._set_attr_from_config_option of <coverage.config.CoverageConfig object at 0x10544f7d0>>]
 ['branch', '->', True]
 ['command_line', '->', None]
 ['concurrency', '->', []]
@@ -12,12 +13,12 @@
 ['config_files_attempted', '->', ['/Users/bdraco/yarl/.coveragerc']]
 ['config_files_read', '->', ['/Users/bdraco/yarl/.coveragerc']]
 ['context', '->', None]
-['copy', '->', <bound method CoverageConfig.copy of <coverage.config.CoverageConfig object at 0x1065a5c70>>]
+['copy', '->', <bound method CoverageConfig.copy of <coverage.config.CoverageConfig object at 0x10544f7d0>>]
 ['cover_pylib', '->', False]
 ['data_file', '->', '/Users/bdraco/yarl/.coverage']
 ['debug', '->', []]
 ['debug_file', '->', None]
-['debug_info', '->', <bound method CoverageConfig.debug_info of <coverage.config.CoverageConfig object at 0x1065a5c70>>]
+['debug_info', '->', <bound method CoverageConfig.debug_info of <coverage.config.CoverageConfig object at 0x10544f7d0>>]
 ['disable_warnings', '->', []]
 ['dynamic_context', '->', None]
 ['exclude_also', '->', []]
@@ -25,10 +26,10 @@
 ['extra_css', '->', None]
 ['fail_under', '->', 0.0]
 ['format', '->', None]
-['from_args', '->', <bound method CoverageConfig.from_args of <coverage.config.CoverageConfig object at 0x1065a5c70>>]
-['from_file', '->', <bound method CoverageConfig.from_file of <coverage.config.CoverageConfig object at 0x1065a5c70>>]
-['get_option', '->', <bound method CoverageConfig.get_option of <coverage.config.CoverageConfig object at 0x1065a5c70>>]
-['get_plugin_options', '->', <bound method CoverageConfig.get_plugin_options of <coverage.config.CoverageConfig object at 0x1065a5c70>>]
+['from_args', '->', <bound method CoverageConfig.from_args of <coverage.config.CoverageConfig object at 0x10544f7d0>>]
+['from_file', '->', <bound method CoverageConfig.from_file of <coverage.config.CoverageConfig object at 0x10544f7d0>>]
+['get_option', '->', <bound method CoverageConfig.get_option of <coverage.config.CoverageConfig object at 0x10544f7d0>>]
+['get_plugin_options', '->', <bound method CoverageConfig.get_plugin_options of <coverage.config.CoverageConfig object at 0x10544f7d0>>]
 ['html_dir', '->', 'htmlcov']
 ['html_skip_covered', '->', None]
 ['html_skip_empty', '->', None]
@@ -45,8 +46,8 @@
 ['paths', '->', {'source': ['yarl', 'tests', '*/lib/pypy*/site-packages/yarl', '*/lib/python*/site-packages/yarl', '*/Lib/site-packages/yarl']}]
 ['plugin_options', '->', {}]
 ['plugins', '->', ['Cython.Coverage']]
-['post_process', '->', <bound method CoverageConfig.post_process of <coverage.config.CoverageConfig object at 0x1065a5c70>>]
-['post_process_file', '->', <bound method CoverageConfig.post_process_file of <coverage.config.CoverageConfig object at 0x1065a5c70>>]
+['post_process', '->', <bound method CoverageConfig.post_process of <coverage.config.CoverageConfig object at 0x10544f7d0>>]
+['post_process_file', '->', <bound method CoverageConfig.post_process_file of <coverage.config.CoverageConfig object at 0x10544f7d0>>]
 ['precision', '->', 0]
 ['relative_files', '->', False]
 ['report_contexts', '->', None]
@@ -54,7 +55,7 @@
 ['report_omit', '->', None]
 ['run_include', '->', []]
 ['run_omit', '->', []]
-['set_option', '->', <bound method CoverageConfig.set_option of <coverage.config.CoverageConfig object at 0x1065a5c70>>]
+['set_option', '->', <bound method CoverageConfig.set_option of <coverage.config.CoverageConfig object at 0x10544f7d0>>]
 ['show_contexts', '->', False]
 ['show_missing', '->', True]
 ['sigterm', '->', False]
@@ -65,4 +66,5 @@
 ['source_pkgs', '->', ['tests', 'yarl']]
 ['timid', '->', False]
 ['xml_output', '->', 'coverage.xml']
-['xml_package_depth', '->', 99]
\ No newline at end of file
+['xml_package_depth', '->', 99]
```

Reply:
Is having `"."` in the report incorrect, though? Codecov v3 was handling that properly, it seems.

Reply:
It looks like they will always get collapsed to `"."` unless there are multiple packages.

Reply:
https://github.com/nedbat/coveragepy/blob/b5408df5c608762d92e90ef7200eb94eb247f0bc/coverage/xmlreport.py#L179

Looks like its always collapsing the test path down so all the test files end up at the top level so codecov looks for `test_xxx.py` instead of `tests/test_xxx.py`

Reply:
Creating `tests/__init__.py` and setting tests to be a package fixes it but it breaks the tests

Reply:
```
['package_name', '.', '/Users/bdraco/yarl/tests/test_cached_property.py']
['package_name', '.', '/Users/bdraco/yarl/tests/test_normalize_path.py']
['package_name', '.', '/Users/bdraco/yarl/tests/test_pickle.py']
['package_name', '.', '/Users/bdraco/yarl/tests/test_quoting.py']
['package_name', '.', '/Users/bdraco/yarl/tests/test_update_query.py']
['package_name', '.', '/Users/bdraco/yarl/tests/test_url.py']
['package_name', '.', '/Users/bdraco/yarl/tests/test_url_build.py']
['package_name', '.', '/Users/bdraco/yarl/tests/test_url_cmp_and_hash.py']
['package_name', '.', '/Users/bdraco/yarl/tests/test_url_parsing.py']
['package_name', '.', '/Users/bdraco/yarl/tests/test_url_query.py']
['package_name', '.', '/Users/bdraco/yarl/tests/test_url_update_netloc.py']
['package_name', 'yarl', '/Users/bdraco/yarl/yarl/__init__.py']
['package_name', 'yarl', '/Users/bdraco/yarl/yarl/_quoting.py']
['package_name', 'yarl', '/Users/bdraco/yarl/yarl/_quoting_c.pyx']
['package_name', 'yarl', '/Users/bdraco/yarl/yarl/_quoting_py.py']
['package_name', 'yarl', '/Users/bdraco/yarl/yarl/_url.py']
```

Reply:
I usually add linters disallowing `tests/__init__.py`. Plus, modern Coverage.py has the distinction of `source` and `source_pkgs` specifically to address the ambiguity.

Reply:
Since `tests` is in `sources` https://github.com/nedbat/coveragepy/blob/b5408df5c608762d92e90ef7200eb94eb247f0bc/coverage/xmlreport.py#L184 will always strip off `tests` so we end up with `test_url_update_netloc.py` instead of `tests/test_url_update_netloc.py` and codecov can't find it

Reply:
@bdraco does `--cov-report=html` also generates something broken? Or is it just XML? Why did Codecov v3 work, then? Own hacks?

Reply:
If what you're saying is correct, we should probably tell Ned about the bug...

Reply:
You gave me an idea.

Reply:
@bdraco try this https://github.com/aio-libs/yarl/pull/1074

Reply:
I'm going to bed now, but feel free to merge it if you think that it is a proper fix.

Reply:
Just saw your comment as I was deep in debugging.  🤦